### ISSUE TITLE
Fails to calculate arguments length, when combining tuple spread with extra arguments

### ISSUE THREAD

**Issue by hpohlmeyer** [microsoft/TypeScript#29248]
**TypeScript Version:**  3.3.0-dev.20190103

**Search Terms:** tuple, array, spread, arguments, length

**Code**

```ts
const three: [number, number, number] = [1, 2, 3];
const two: [number, number] = [1, 2];
const one: [number] = [1]; 

const foo = (a: number, b: number, c: number) => undefined;

// EXPECTED: correctly realizes there are three args
foo(...three);

// EXPECTED: correctly realizes there are two args
// error: Expected 3 arguments, but got 2.
foo(...two);

// EXPECTED: correctly realizes there is one arg
// error: Expected 3 arguments, but got 1.
foo(...one);

// UNEXPECTED: does not realize there are three args
// error: Expected 3 arguments, but got 1 or more.
foo(...two, 2);

// UNEXPECTED: does not realize there are three args
// error: Expected 3 arguments, but got 1 or more.
foo(...one, 2, 3);
```

**Expected behavior:**
Correctly calculate the count of the arguments passed into `foo`. Add the additional arguments to the length of the tuple. Basically the last two examples should not throw an error.

**Actual behavior:**
When adding additional arguments after the spreaded tuple, TS fails to calculate the total amount of arguments end expects the amount of the additional arguments "+ more".

**Playground Link:** [Link](http://www.typescriptlang.org/play/#src=const%20three%3A%20%5Bnumber%2C%20number%2C%20number%5D%20%3D%20%5B1%2C%202%2C%203%5D%3B%0D%0Aconst%20two%3A%20%5Bnumber%2C%20number%5D%20%3D%20%5B1%2C%202%5D%3B%0D%0Aconst%20one%3A%20%5Bnumber%5D%20%3D%20%5B1%5D%3B%20%0D%0A%0D%0Aconst%20foo%20%3D%20(a%3A%20number%2C%20b%3A%20number%2C%20c%3A%20number)%20%3D%3E%20undefined%3B%0D%0A%0D%0A%2F%2F%20EXPECTED%3A%20correctly%20realizes%20there%20are%20three%20args%0D%0Afoo(...three)%3B%0D%0A%0D%0A%2F%2F%20EXPECTED%3A%20correctly%20realizes%20there%20are%20two%20args%0D%0Afoo(...two)%3B%0D%0A%0D%0A%2F%2F%20EXPECTED%3A%20correctly%20realizes%20there%20is%20one%20arg%0D%0Afoo(...one)%3B%0D%0A%0D%0A%2F%2F%20UNEXPECTED%3A%20does%20not%20realize%20there%20are%20three%20args%0D%0Afoo(...two%2C%203)%3B%0D%0A%0D%0A%2F%2F%20UNEXPECTED%3A%20does%20not%20realize%20there%20are%20three%20args%0D%0Afoo(...one%2C%202%2C%203)%3B%0D%0A)

**Related Issues:**

**Comment by ahejlsberg** [comment]
I'm changing this to a suggestion as the current behavior is actually intended. From the description in #24897:

> When a function call includes a spread expression of a tuple type _as the last argument_, the spread expression corresponds to a sequence of discrete arguments of the tuple element types.

A spread expression in any position but the last is always processed as a spread of an array type, i.e. we consider it to contribute an unknown number of arguments.

A key issue with this suggestion is whether we're willing to trust that a fixed-length tuple type always has exactly the declared fixed length. We do the best we can to check it, but it is obviously easy to circumvent.

**Comment by hpohlmeyer** [comment]
I was just going through my open bug reports and looked at my playground link.  
The behavior seems to have changed with TS 4.0 and [variadic tuple types](https://devblogs.microsoft.com/typescript/announcing-typescript-4-0/#variadic-tuple-types).

I guess it can be closed than.
